tests/test_external_strategies.py

```diff
@@ -0,0 +1,119 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+from unittest.mock import MagicMock, Mock
+
+from d3a.models.area import Area
+from d3a.models.strategy.external_strategies.load import LoadForecastExternalStrategy
+from d3a.models.strategy.external_strategies.pv import PVForecastExternalStrategy
+from d3a_interface.constants_limits import GlobalConfig
+from pendulum import now, Duration
+from pytest import mark
+
+
+def ext_strategy_fixture(strategy):
```

Comment:
I'm a bit confused: pytest fixtures are defined using the [`@pytest.fixture`](https://docs.pytest.org/en/6.2.x/fixture.html) decorator, but I don't see it here. This looks like a normal function (so we could use a naming like `generate_strategy` or similar.

A more proper alternative might be using `@pytest.fixture` with the `indirect` flag (so that you could still pass the `strategy` argument):
- [Indirect parametrization](https://docs.pytest.org/en/latest/example/parametrize.html#indirect-parametrization)
- [Pass a parameter to a fixture function](https://stackoverflow.com/a/33879151/1916588)

Reply:
I was expecting this comment. The problem is that you can not pass pytest.fixtures into `@mark.parametrize`. So i just "mocked" the fixture.

Reply:
I think this is what the second link is doing: it uses just one fixture (with a parameter which would be the Strategy in our case). Or am I missing something?

Reply:
I tried it with 
```python
from pytest import mark, fixture


@fixture
def ext_strategy_fixture(strategy):
    config = Mock()
    config.slot_length = duration(minutes=15)
    config.tick_length = duration(seconds=15)
    config.ticks_per_slot = 90
    config.start_date = GlobalConfig.start_date
    config.grid_fee_type = ConstSettings.IAASettings.GRID_FEE_TYPE
    config.end_date = GlobalConfig.start_date + duration(days=1)
    config.market_count = 1
    area = Area(name="forecast_pv", config=config, strategy=strategy,
                external_connection_available=True)
    parent = Area(name="parent_area", children=[area], config=config)
    parent.activate()
    area.activate()
    strategy.connected = True
    market = MagicMock()
    market.time_slot = GlobalConfig.start_date
    return strategy


class TestPVForecastExternalStrategy:

    @mark.parametrize("ext_strategy_fixture", [LoadForecastExternalStrategy(),
                                               PVForecastExternalStrategy()], indirect=True)
    def test_event_market_cycle_calls_energy_update_methods(self, ext_strategy_fixture):
        ext_strategy_fixture.energy_forecast_buffer = {now(): 1}
        ext_strategy_fixture.energy_measurement_buffer = {now(): 1}
        ext_strategy_fixture.update_energy_forecast = Mock()
        ext_strategy_fixture.update_energy_measurement = Mock()
        ext_strategy_fixture.event_market_cycle()
        ext_strategy_fixture.update_energy_forecast.assert_called_once()
        ext_strategy_fixture.update_energy_measurement.assert_called_once()
        assert ext_strategy_fixture.energy_forecast_buffer == {}
        assert ext_strategy_fixture.energy_measurement_buffer == {}
```

But got 
```
test setup failed
file /Users/hannesd/d3a/d3a/tests/test_external_strategies.py, line 52
      @mark.parametrize("ext_strategy_fixture", [LoadForecastExternalStrategy(),
                                     PVForecastExternalStrategy()], indirect=["ext_strategy_fixture"])
      def test_event_market_cycle_calls_energy_update_methods(self, ext_strategy_fixture):
file /Users/hannesd/d3a/d3a/tests/test_external_strategies.py, line 29
  @fixture
  def ext_strategy_fixture(strategy):
E       fixture 'strategy' not found
>       available fixtures: cache, called, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, ext_strategy_fixture, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, requests_mock, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```
I dont understand really (not even the stackoverflow post. Sorry. Can you help?

Reply:
I think all you have to do is change this `def ext_strategy_fixture(strategy):` to this `def ext_strategy_fixture():` . Fixture function arguments should be either other fixtures, or one of the predefined fixtures from pytest (e.g. `cache`, `called` ... ). `strategy` is not a symbol in this file as far as I see.

Reply:
> I think all you have to do is change this `def ext_strategy_fixture(strategy):` to this `def ext_strategy_fixture():` . Fixture function arguments should be either other fixtures, or one of the predefined fixtures from pytest (e.g. `cache`, `called` ... ). `strategy` is not a symbol in this file as far as I see.

But how do I pass the strategy in there?

Reply:
I just noticed that you need to use the strategy inside the fixture. In order to do that, you have 2 options: 

1. construct the strategy class inside the fixture. Bad, because it does not allow parameterization of the test (as far as I saw you use the constructed strategy object as a test parameter, therefore this will not work easily)
2. do not use a module-scope fixture for this, but use a class method, that will be called right at the start of the test and will set up the strategy the way you want it. 
3. Create your own fixture decorator that would receive the output of the parameterized decorator and set up the already constructed strategy. More complex, and without much gain in readability of the test. 

I would go with option #2

Reply:
@hannesdiedrich in order to pass and use the parameterized strategy with indirect parameterization, you need to use the `request` object. I used the snippet you posted above and just changed the first lines:
```python
@fixture
def ext_strategy_fixture(request):
    strategy = request.param
    # ...
```
It should work.

Reply:
You can find more information on the documentation:
- [`parametrize`](https://docs.pytest.org/en/latest/reference/reference.html?highlight=request#pytest.Metafunc.parametrize)

Specifically these lines regarding the `indirect` parameter:
> **indirect** (_Union[bool, Sequence[str]]_) – A list of arguments’ names (subset of argnames) or a boolean. If True the list contains all names from the argnames. Each argvalue corresponding to an argname in this list will be passed as request.param to its respective argname fixture function so that it can perform more expensive setups during the setup phase of a test rather than at collection time.

Reply:
Thanks a lot for pointing me to the correct point in the documentation. I need to learn to be more patient here and not to stop reading when I think I found the solution and then wonder why it does not work.
I adapted your proposal here https://github.com/gridsingularity/d3a/pull/1200/commits/87a5e4f1df6a16aa130f3e3dd1a4ae18cec980bb